pkg/oauth/discovery/discovery.go

```diff
@@ -0,0 +1,51 @@
+package discovery
+
+import (
+	"github.com/RangelReale/osin"
+	"github.com/openshift/origin/pkg/authorization/authorizer/scope"
+	"github.com/openshift/origin/pkg/oauth/server/osinserver"
+)
+
+// OauthAuthorizationServerMetadata holds OAuth 2.0 Authorization Server Metadata used for discovery
+// https://tools.ietf.org/id/draft-ietf-oauth-discovery-04.html#rfc.section.2
+type OauthAuthorizationServerMetadata struct {
+	// The authorization server's issuer identifier, which is a URL that uses the https scheme and has no query or fragment components.
+	// This is the location where .well-known RFC 5785 [RFC5785] resources containing information about the authorization server are published.
+	Issuer string `json:"issuer"`
+
+	// URL of the authorization server's authorization endpoint [RFC6749].
+	AuthorizationEndpoint string `json:"authorization_endpoint"`
+
+	// URL of the authorization server's token endpoint [RFC6749].
+	TokenEndpoint string `json:"token_endpoint"`
+
+	// JSON array containing a list of the OAuth 2.0 [RFC6749] scope values that this authorization server supports.
+	// Servers MAY choose not to advertise some supported scope values even when this parameter is used.
+	ScopesSupported []string `json:"scopes_supported"`
+
+	// JSON array containing a list of the OAuth 2.0 response_type values that this authorization server supports.
+	// The array values used are the same as those used with the response_types parameter defined by "OAuth 2.0 Dynamic Client Registration Protocol" [RFC7591].
+	ResponseTypesSupported osin.AllowedAuthorizeType `json:"response_types_supported"`
+
+	// JSON array containing a list of the OAuth 2.0 grant type values that this authorization server supports.
+	// The array values used are the same as those used with the grant_types parameter defined by "OAuth 2.0 Dynamic Client Registration Protocol" [RFC7591].
+	GrantTypesSupported osin.AllowedAccessType `json:"grant_types_supported"`
+}
+
+func Get(masterPublicURL, authorizeURL, tokenURL string) OauthAuthorizationServerMetadata {
+	config := osinserver.NewDefaultServerConfig()
+	return OauthAuthorizationServerMetadata{
+		Issuer:                masterPublicURL,
+		AuthorizationEndpoint: authorizeURL,
+		TokenEndpoint:         tokenURL,
+		ScopesSupported: []string{ // Note: this list is incomplete, which is allowed per the draft spec
```

Comment:
Right, the remaining role scope is varying, since it contains actual role name and namespace. Well, we're complaint with the spec, since it allows not to advertise all. Thanks for the info